<html><title>PseudoBase information</title>
{% extends "base.html" %}

{% block page_title %}Information{% endblock %}

{% block links %}{% endblock %}


{% block header %}

  <div class = "col-md-1">
  </div> 
  <div class = "col-md-10">
    <center>
    
    <h1>PseudoBase Information</h1>
    </center>
  </div>
  <div class = "col-md-1">
  </div>
    {% endblock header %}
    
    
    {% block content %}

<ul class="nav nav-tabs" id="myTab" role="tablist">
  <li class="nav-item">
    <a class="nav-link active" id="source-tab" data-toggle="tab" href="#source" role="tab" aria-controls="source" aria-selected="false">Source Data</a>
  </li>
  <li class="nav-item">
    <a class="nav-link" id="lines-tab" data-toggle="tab" href="#lines" role="tab" aria-controls="lines" aria-selected="true">Lines available</a>
  </li>

  <li class="nav-item">
    <a class="nav-link" id="user-tab" data-toggle="tab" href="#user" role="tab" aria-controls="user" aria-selected="false">User Interface</a>
  </li>
</ul>
<div class="tab-content" id="myTabContent">
  <div class="tab-pane fade" id="lines" role="tabpanel" aria-labelledby="lines-tab">
<div class = "col-md-1">
</div>
<div class="col-md-10 mt-4 border border-primary" style="background-color:white";>
    <h4 class="mt-2">Lines available (not all lines available for all queries)</h4>
    <div class = "mt-4"> <!-- previously used card-columns. Now replaced with manually placed cards-->
       <div class = "row">
        {% for species in species_tab %}
             {%  if species.name == "D.pseudoobscura" or species.name == "D.pseudoobscura bogotana" or species.name == "D.miranda" %}
               <!-- Note: best to replace above line with better logic to arrange the species cards-->
               <div class="col">
             {%  endif %}
                <div class = "card">
                  <div class = "card-header bg-info">
                   <h5 class="card-title"><b><i>{{ species.name }}</i></b></h5>
                   <h6 class = "card-subtitle mb-2 text-muted">{{ species.num_lines }}
                       {% if species.num_lines == 1 %}
                          line
                       {% else %}
                          lines
                       {% endif %}
                   </h6>
                  </div>
                     <!--ul class="list-unstyled"-->
                     <div class = "list-group">
                         {% for line in species.lines %}

                                <div class = "list-group-item">
                                    <!--li class="text-left"-->
                                    <span class = "font-weight-bold">{{ line.name }}</span>
                                    {%  if line.formatted_year != '' %}
                                            <br>
                                            <span class = "small">
                                            {{  line.formatted_year }}
                                            </span>

                                    {%  endif %}
                               {%  if line.formatted_info_dict %}
                                            <br>
                                            <span class = "small">
                                            {%  for key,values in line.formatted_info_dict.items %}
                                             {{  key }} : {{  values }}
                                            {%  endfor %}
                                            </span>

                                    {%  endif %}

                                </div>

                         {% endfor %}
                     </div>

                </div>

             {%  if species.name == "D.pseudoobscura" or species.name == "D.persimilis" or species.name == "D.lowei" %}
                 <!-- Note: best to replace above line with better logic to arrange the species cards-->
               </div>
             {%  endif %}
        {% endfor %}
       </div>
    </div>



</div>
<div class = "col-md-1">
</div>
     </div>
  <div class="tab-pane fade  show active" id="source" role="tabpanel" aria-labelledby="source-tab">
<div class = "col-md-1">
</div>
<div class="col-md-10 mt-4 border border-primary" style="background-color:white";>


    <p>
    <h4>Sequencing Information and Variant Calling</h4>
<p class="text-left">
    PseudoBase uses whole genome paired-end Illumina sequencing from multiple laboratory groups and experiments (associated publications include: Fuller, Leonard, Young, Schaeffer, & Phadnis, 2018; Korunes et al., 2019; McGaugh et al., 2012; McGaugh & Noor, 2012; Samuk, Manzano-Winkler, Ritz, & Noor, 2020). Raw sequencing data and associated details are available on the NCBI Short Read Archive under the sample accessions provided in the "lines availabel" tab.  <p class="text-left">A brief note on naming conventions: there are two named subspecies of <i>D. pseudoobscura—D. pseudoobscura pseudoobscura</i> and <i>D. pseudoobscura bogotana.</i> In PseudoBase and in our associated paper, we use <i>D. pseudoobscura</i> to refer to both subspecies. We specify <i>D. pseudoobscura pseudoobscura</i> or <i>D. pseudoobscura bogotana</i> when we are specifically referring to only one of the two subspecies</p><p class="text-left">The pipeline used for genome alignment and variant calling is available on GitHub (<a href="https://github.com/kkorunes/PseudobaseScripts">https://github.com/kkorunes/PseudobaseScripts</a>). <br>
    In summary, we first used BWA-0.7.17 (Li & Durbin 2009) to align all sequences to the D. pseudoobscura genome assembly, obtained from FlyBase (Dpse_3.04: GCA_000001765.2; Thurmond et al. 2019). We used Picard to mark adapters and duplicates that might introduce bias from data generation steps such as PCR amplification (<a href="http://broadinstitute.github.io/picard/">http://broadinstitute.github.io/picard/</a>). Variants were then called and filtered using GATK v4.1.1 (McKenna et al. 2010; Van der Auwera et al. 2013). We filtered SNPs and INDELs separately, according to the hard filtering recommendations provide by GATK. Specifically, we excluded SNPs with QualByDepth (QD) < 2.0 , FisherStrand (FS) > 60, and StrandOddsRatio (SOR) > 3.0, MQ < 40, MQRankSum < -12.5, ReadPosRankSum < -8. INDELs were filtered to exclude variants with QualByDepth (QD) < 2.0 , FisherStrand (FS) > 200, and StrandOddsRatio (SOR) > 10.0, ReadPosRankSum < 20.</p>
</p>
    <h4>Citations:</h4>
<ul>
<li>
<p class="text-left">
    Buels, R., Yao, E., Diesh, C. M., Hayes, R. D., Munoz-Torres, M., Helt, G., … Holmes, I. H. (2016). JBrowse: a dynamic web platform for genome visualization and analysis. <i>Genome Biology</i>, 17(66). doi: 10.1186/s13059-016-0924-1
</p>
</li>
<li>
<p class="text-left">
    Fuller, Z. L., Leonard, C. J., Young, R. E., Schaeffer, S. W., & Phadnis, N. (2018). Ancestral polymorphisms explain the role of chromosomal inversions in speciation. PLoS Genetics, 14(7), e1007526. doi: 10.1371/journal.pgen.1007526
</p>
</li>
<li>
<p class="text-left">

    Korunes, K. L., Machado, C. A., & Noor, M. A. (2019). Inversions shape the divergence of Drosophila pseudoobscura and D. persimilis on multiple timescales. BioRxiv, 842047. doi: 10.1101/842047
</p>
</li>
<li>
<p class="text-left">
    Li H, Durbin R (2009) Fast and accurate short read alignment with Burrows-Wheeler Transform. <i>Bioinformatics,</i> 25, 1754–60. doi:10.1093/bioinformatics/btp324
</p>
</li>


<li>
<p class="text-left">
McGaugh, S. E., Heil, C. S. S., Manzano-Winkler, B., Loewe, L., Goldstein, S., Himmel, T. L., & Noor, M. A. F. (2012). Recombination modulates how selection affects linked sites in Drosophila. PLoS Biology, 10(11), e1001422. doi: 10.1371/journal.pbio.1001422
</p>
</li>
<li>
<p class="text-left">
McGaugh, S. E., & Noor, M. A. F. (2012). Genomic impacts of chromosomal inversions in parapatric Drosophila species. Philosophical Transactions of the Royal Society B: Biological Sciences, 367(1587), 422–429. doi: 10.1098/rstb.2011.0250
</p>
</li>
<li>
<p class="text-left">
    McKenna A, Hanna M, Banks E et al. (2010) The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. <i>Genome Research,</i> 20, 1297–303. doi:10.1101/gr.107524.110
</p>
</li>
<li>
<p class="text-left">
Samuk, K., Manzano-Winkler, B., Ritz, K. R., & Noor, M. A. F. (2020). Natural selection shapes variation in genome-wide recombination rate in Drosophila pseudoobscura. Current Biology, 30(8), 1517-1528.E6. doi: 10.1016/j.cub.2020.03.0
</p>
</li>
<li>
<p class="text-left">
    Thurmond J, Goodman JL, Strelets VB et al. (2019) FlyBase 2.0: The next generation. <i>Nucleic Acids Research,</i> 47, D759–D765. doi:10.1093/nar/gk
</p>
</li>
<li>
<p class="text-left">
    Van der Auwera GA, Carneiro MO, Hartl C <i>et al.</i> (2013) From FastQ data to high-confidence variant calls: The Genome Analysis Toolkit best practices pipeline. <i>Current Protocols in Bioinformatics,</i> 43, 11.10.1-11.10.33. doi:10.1002/0471250953.bi1110s43
</p>
</li>

</ul>
</div>
<div class = "col-md-1">
</div>

  </div>
  <div class="tab-pane fade" id="user" role="tabpanel" aria-labelledby="user-tab">
    <div class = "col-md-1">
    </div>
    <div class="col-md-10 mt-4 border border-primary" style="background-color:white";>
        <h4 class="mt-4">Pseudobase Homepage</h4>
        <p class="text-left">The PseudoBase <a href = "{%  url 'index' %}">homepage</a> allows the user to query by gene (or genes if the user uploads a batch query) or by chromosomal region. By selecting one or more species of interest, the user can either generate a FASTA-formatted alignment or navigate to the JBrowse interface (Buels et al., 2016). The JBrowse interface can also be reached directly, using the “<a href = "{%  url 'browse' %}">Browse</a>” tab.</p>

        <h4>Embedded JBrowse</h4>
        <p class = "text-left">The embedded JBrowse interface allows for browsing of specific genes/regions. All strains imported into PseudoBase are automatically made available for browsing within JBrowse, and the user can view or hide strains by checking/unchecking boxes in the “Available Tracks” panel to the left of the JBrowse viewer. JBrowse allows the user to visualize SNPs and indels (tracks with the “I/D” prefix) specific to each selected track. Clicking on any of the displayed variants brings up further details, such as the specific allele and its attributes (e.g., sequencing depth).</p>

        <p class = "text-left">The top of the JBrowse interface includes arrows to navigate along the selected chromosome, -/+ options to zoom out or in, a chromosome dropdown menu to jump to a different chromosome, and a chromosome coordinate entry box to jump to a different region of the selected chromosome.</p>
        <p class="text-left">For the genomic region selected in the JBrowse viewer, the reference sequence and accompanying annotations are pulled from FlyBase (<a href="ftp://ftp.flybase.net/genomes/Drosophila_pseudoobscura/dpse_r3.04_FB2018_05/gff/dpse-all-3.04.gff.gz">ftp://ftp.flybase.net/genomes/Drosophila_pseudoobscura/dpse_r3.04_FB2018_05/gff/dpse-all-3.04.gff.gz</a>). The displayed features include genes, coding sequences (CDSs), exons, introns, untranslated regions (5’ and 3’ UTRs), mRNA, ncRNA, orthologous regions, “orthologous to" annotations, proteins, and syntenic_regions. See the <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#FASTA_files">FlyBase documentation</a> for descriptions of these data types. Clicking on any of these features brings up detailed information, including coordinates, the feature length, any aliases, the full nucleotide sequence, and the nucleotide sequence of each subfeature (e.g., introns).</>
    </div>
    <div class = "col-md-1">
    </div>
  </div>
</div>


{% endblock content %}