<html><title>Pseudobase information</title>
{% extends "base.html" %}

{% block page_title %}Information{% endblock %}

{% block links %}{% endblock %}


{% block header %}

  <div class = "col-md-1">
  </div> 
  <div class = "col-md-10">
    <center>
    
    <h1>Pseudobase Information</h1>
    </center>
  </div>
  <div class = "col-md-1">
  </div>
    {% endblock header %}
    
    
    {% block content %}
<div class = "col-md-1">
</div>   
<div class="col-md-10 border border-primary" style="background-color:white";>    
    <h4 class="mt-2">Lines available (not all lines available for all queries)</h4>
    <div class = "card-columns mt-4">  
        {% for species in species_tab %}
            <div class = "card">
              <div class = "card-header bg-light">
               <h5 class="card-title"><b><i>{{ species.name }}</i></b></h5>
               <h6 class = "card-subtitle mb-2 text-muted">{{ species.num_lines }}
                   {% if species.num_lines == 1 %}
                      line
                   {% else %}   
                      lines
                   {% endif %}   
               </h6>
              </div>
                 <!--ul class="list-unstyled"-->
                 <div class = "list-group">
                     {% for line in species.lines %}

                            <div class = "list-group-item">
                                <!--li class="text-left"-->
                                <span class = "font-weight-bold">{{ line.name }}</span>
                                {%  if line.formatted_year != '' %}
                                        <br>
                                        <span class = "small">
                                        {{  line.formatted_year }}
                                        </span>

                                {%  endif %}
                           {%  if line.formatted_info_dict %}
                                        <br>
                                        <span class = "small">
                                        {%  for key,values in line.formatted_info_dict.items %}
                                         {{  key }} : {{  values }}
                                        {%  endfor %}
                                        </span>

                                {%  endif %}

                            </div>

                     {% endfor %}
                 </div>

            </div>    
        {% endfor %}   
    </div>     
    
    
    <p>
    <h4>Sequencing Information</h4></p>
    <p>
    For each line, DNA from 15-20 virgin female inbred flies was extracted using the GentraSystems PureGene DNA preparation and sequenced with Illumina Genome Analyzer IIx at The University of North Carolina-Chapel Hill High Throughput Sequencing Facility. Illumina reads were aligned to the <i>Drosophila pseudoobscura</i> reference genome v2.9 using bwa-0.5.5 (default alignment settings were used except the maximum number of gap extensions was set to 4). Consensus assemblies for each resequenced line were generated using the bwa alignments and Samtools 0.1.6 pileup. Default pileup settings were used except the theta parameter (error dependency coefficient) in the maq consensus calling model was set to 0.9, and the number of haplotypes in the sample was set to one because each resequenced line was inbred.
    </p>
    <p>
        The raw traces used to generate the alignments here are available from the NIH NCBI Short Read Archive under accession SRP007802. Additional related sequences are available under PRJNA277849.
    </p>
</div>
<div class = "col-md-1">
</div>   

{% endblock content %}